commands/runserver/runserver.drush.inc

```diff
@@ -134,6 +135,43 @@ function drush_core_runserver($uri = NULL) {
   else {
     $browse = drush_url($path);
   }
+  // Special checking for 'default' on localhost
+  if (preg_match('#://default/#', $browse) && drush_is_local_host($uri['host'])) {
+    $server = $uri["host"];
+    if (!empty($uri["port"])) {
+      $server .= ':' . $uri["port"];
+    }
+    $browse = str_replace('://default/', '://' . $server . '/', $browse);
+  }
+  // If the user specified --cache-uri, then create a drushrc.php so that
+  // future calls to drush uli will work seemlessly.
+  if (drush_get_option('cache-uri')) {
+    $dir = DRUPAL_ROOT . '/drush';
+    $target_file = $dir . '/drushrc.php';
+    $create_file = TRUE;
+    if (file_exists($target_file)) {
+      // Don't bother to ask with --use-existing; just
+      // continue.
+      if (drush_get_option('use-existing', FALSE)) {
+        $create_file = FALSE;
+      }
+      else {
+        $create_file = drush_confirm(dt('!target already exists. Overwrite?', array('!target' => $target_file)));
+      }
+    }
+    $content = <<<EOT
+<?php
+
+// Inserted by `drush runserver`.  This allows `drush user-login`
+// and similar commands to work seemlessly.  Remove if using 
+// Drupal multisite feature.
+\$options['uri'] = "$server";
+EOT;
+    if ($create_file) {
+      drush_mkdir($dir);
+      file_put_contents($target_file, $content);
```

Comment:
The purpose of setting $options['uri'] here is to literally set the --uri option, so that it is correct in whatever context it might be needed.  If we used the Drush cache API, then we'd need to also add special-checking to the user-login command that used the cache value instead of --uri.  Then we also have to make sure that --uri overrides the cache value iff it was user-specified, and doesn't if it is a default.